### PR TITLE
New authentication method for COSU using OAuth

### PR DESCRIPTION
**Pull request by janakamarasena**
Added new authentication method for COSU using OAuth to seamlessly enroll without the use of certificates.

### CODE REVIEW

**Review by chaturadilan — APPROVED** [review]

File: client/client/src/main/java/org/wso2/iot/agent/activities/AuthenticationActivity.java [hunk 1]
```diff
@@ -529,8 +539,12 @@ private void initializeIDPLib(String clientKey, String clientSecret) {
 			info.setPassword(passwordVal);
 			info.setTokenEndPoint(serverURL);
 
+			if (adminAccessToken != null) {
+				info.setAdminAccessToken(adminAccessToken);
+			}
+
 			//adding device-specific scope
-			String deviceScope = "deivce_" + deviceInfo.getDeviceId();
+			String deviceScope = "device_" + deviceInfo.getDeviceId();
```
Comment (malithie) [inline client/client/src/main/java/org/wso2/iot/agent/activities/AuthenticationActivity.java:547]: This prefix should probably have a constant and it should have been reused

File: client/client/src/main/java/org/wso2/iot/agent/services/AgentDeviceAdminReceiver.java [hunk 2]
```diff
@@ -184,8 +186,21 @@ public void onProfileProvisioningComplete(Context context, Intent intent) {
...
                 }
             }
-
             Intent launch = new Intent(context, SplashActivity.class);
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                PersistableBundle persistableBundle = intent.getParcelableExtra(
+                        DevicePolicyManager.EXTRA_PROVISIONING_ADMIN_EXTRAS_BUNDLE);
+                if (persistableBundle != null) {
+                    if (org.wso2.iot.agent.proxy.utils.Constants.Authenticator.AUTHENTICATOR_IN_USE.
+                            equals(org.wso2.iot.agent.proxy.utils.Constants.Authenticator.OAUTH_AUTHENTICATOR)) {
```
Comment (malithie) [inline client/client/src/main/java/org/wso2/iot/agent/services/AgentDeviceAdminReceiver.java:195]: What's the requirement to check if two constants are equal

File: client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java [hunk 3]
```diff
@@ -83,6 +83,32 @@ public void getClientCredentials(String username, String password, ServerConfig
...
+     * This method is used to register an oauth application in the backend using an access token in
+     * COSU mode.
+     *
+     * @param apiRegistrationProfile Payload of the register request.
+     * @param utils Server configurations.
+     *
+     * @return returns consumer key and consumer secret if success. Else returns null
+     *         if it fails to register.
+     * @throws AndroidAgentException
+     */
```
Comment (malithie) [inline client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java:96]: There are several params not explained in the doc comment of this public method

File: client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java [hunk 4]
```diff
@@ -218,6 +244,112 @@ public void onErrorResponse(VolleyError error) {
...
+        int requestMethod = 0;
+        org.wso2.iot.agent.proxy.utils.Constants.HTTP_METHODS httpMethod = endPointInfo.getHttpMethod();
+        switch (httpMethod) {
+            case POST:
+                requestMethod = Request.Method.POST;
+                break;
+            case DELETE:
+                requestMethod = Request.Method.DELETE;
+                break;
+        }
```
Comment (malithie) [inline client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java:263]: Above could have been broken to separate method for clarity

File: client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java [hunk 5]
```diff
@@ -218,6 +244,112 @@ public void onErrorResponse(VolleyError error) {
...
+                    DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
+        } catch (JSONException e) {
+            Log.e(TAG, "Failed to parse request JSON", e);
+        }
+        request.setRetryPolicy(new DefaultRetryPolicy(
+                org.wso2.iot.agent.proxy.utils.Constants.HttpClient.DEFAULT_TIME_OUT,
+                DefaultRetryPolicy.DEFAULT_MAX_RETRIES,
+                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
+        queue.add(request);
+    }
```
Comment (malithie) [inline client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java:351]: This method could have been broken for readability

File: client/client/src/main/java/org/wso2/iot/agent/services/AgentDeviceAdminReceiver.java (hunk 2 above)
Comment (janakamarasena) [reply client/client/src/main/java/org/wso2/iot/agent/services/AgentDeviceAdminReceiver.java:195]: At the build time, it is possible to change the constant value using the build config.

File: client/client/src/main/java/org/wso2/iot/agent/activities/AuthenticationActivity.java (hunk 1 above)
Comment (janakamarasena) [reply client/client/src/main/java/org/wso2/iot/agent/activities/AuthenticationActivity.java:547]: Fixed with https://github.com/wso2/cdmf-agent-android/pull/199

File: client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java (hunk 3 above)
Comment (janakamarasena) [reply client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java:96]: Fixed with https://github.com/wso2/cdmf-agent-android/pull/199

File: client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java (hunk 4 above)
Comment (janakamarasena) [reply client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java:263]: Fixed with https://github.com/wso2/cdmf-agent-android/pull/199

File: client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java (hunk 5 above)
Comment (janakamarasena) [reply client/client/src/main/java/org/wso2/iot/agent/services/DynamicClientManager.java:351]: Fixed with https://github.com/wso2/cdmf-agent-android/pull/199

**Pull request merged**